### PR TITLE
add support of impl block for doctest into runnables

### LINKED ISSUE #6356: "Run Doctest" lens only available for functions
Sometimes it makes sense to write a doctest for a [struct](https://docs.rs/tune/0.18.0/src/tune/ratio.rs.html#10-43) or [trait impl](https://docs.rs/tune/0.18.0/src/tune/tuning.rs.html#18-31).

Unfortunately, for those positions no "Run Doctest" lens is shown. It would be great it was available for any possible doctest position! :smiley:

### PR DESCRIPTION
**Pull request by bnjjj**
close #6356

### CODE REVIEW

**Comment by matklad** [comment]
could you verify that the doctest also is run if this is a trait impl? Are there any other special cases?

**Comment by bnjjj** [comment]
I added test for trait impl. But I was not able to find documentation to fit a trait impl block for TestId. I tried `Trait::Struct` but it didn't work. if you have any info about this I'm interested

**Comment by bnjjj** [comment]
Any idea ?

**Comment by matklad** [comment]
Sorry I dont' understand, could you rephrase the queston?

Are you asking what string should we pass on the command like to cargo to make it actually execute the specific test?

**Comment by bnjjj** [comment]
Yes sorry, exactly for example in my test here https://github.com/rust-analyzer/rust-analyzer/pull/8952/files#diff-7130573530754c75b24220435cc423a6e191799e016fed3605a7ae67ab99a7ecR785 I use the struct name as the TestId for an impl Trait for StructName. But doing this doesn't trigger only the doc test for this impl trait block but also for the `impl StructName` without the trait.

**Comment by matklad** [comment]
Seems like it's just impossible to do this unambiguously?

```rust
/// ```
/// assert_eq!(2 + 2, 4)
/// ```
struct S;

/// ```
/// assert_eq!(2 + 2, 4)
/// ```
trait T {
    /// ```
    /// assert_eq!(2 + 2, 4)
    /// ```
    fn t() {}
}

/// ```
/// assert_eq!(2 + 2, 4)
/// ```
impl T for S {
    /// ```
    /// assert_eq!(2 + 2, 4)
    /// ```
    fn t() {}
}

/// ```
/// assert_eq!(2 + 2, 4)
/// ```
impl S {
    /// ```
    /// assert_eq!(2 + 2, 4)
    /// ```
    fn s() {}
}
```

```
λ cargo test -q --doc -j 1
running 7 tests
test src/lib.rs - T (line 6) ... ok
test src/lib.rs - S::s (line 30) ... ok
test src/lib.rs - T::t (line 10) ... ok
test src/lib.rs - S (line 1) ... ok
test src/lib.rs - S (line 16) ... ok
test src/lib.rs - S::t (line 20) ... ok
test src/lib.rs - S (line 26) ... ok
```

Couple of names there match exactly. 

@jyn514 am I correct that the paths printed by doctests are exactly the names of the tests, and we can't distinguish between

```
test src/lib.rs - S (line 16) ... ok
test src/lib.rs - S (line 26) ... ok
```

and specify some argument to run only one of these two tests? Is there canonical doc for this?

**Comment by jyn514** [comment]
I am not aware of a way to distinguish them, nor a canonical doc. The relevant code is in src/librustdoc/doctest.rs, I don't have time to help other than that.

**Comment by bnjjj** [comment]
If there is no other way then @matklad it's ready to merge
